### PR TITLE
[Metrics V2] Query processor handle no join strategy

### LINKED ISSUE #42462: [BE] [QP, MBQL lib] Allow using compatible metrics for the source table

### LINKED ISSUE #42667: [BE] [QP] Allow referencing compatible metrics in queries without joins (previous behavior)

### PR DESCRIPTION
**Pull request by snoe**
Fixes #42667

Make QP preprocessor handle queries referencing metrics in the style of legacy metrics. Enforce various compatibility checks.

### CODE REVIEW

**Review by metamben — CHANGES_REQUESTED** [review]
LGTM, but I have a few questions. Also, I would prefer to have test cases explicitly testing and named for the multi-stage metric execution. I would like to see test showing what happens when expression names in a metric clash with the expression names in the host query.

File: src/metabase/query_processor/middleware/metrics.clj [hunk 1]
```diff
@@ -10,33 +12,68 @@
...
+    (->> metric-ids
+         (lib.metadata/bulk-metadata-or-throw query :metadata/card)
+         (into {}
+               (map (juxt :id
+                          (fn [card-metadata]
+                            (let [metric-query (lib.convert/->pMBQL
+                                                 ((requiring-resolve 'metabase.query-processor.preprocess/preprocess)
+                                                  (lib/query query (:dataset-query card-metadata))))]
+                              {:query metric-query
+                               :aggregation (first (lib/aggregations metric-query))
```
[the remark this answers is not in the document]
Comment (snoe) [reply src/metabase/query_processor/middleware/metrics.clj]: I'm not sure it should be possible.

File: src/metabase/query_processor/middleware/metrics.clj [hunk 2]
```diff
@@ -10,111 +12,160 @@
...
+                               :name (:name card-metadata)})))))
+         not-empty)))
+
+(defn- expression-with-name-from-source
+  [query [_ {:lib/keys [expression-name]} :as expression]]
+  (lib/expression query 0 expression-name expression))
```
[the remark this answers is not in the document]
Comment (snoe) [reply src/metabase/query_processor/middleware/metrics.clj:49]: As per slack convo, I'll address this in a followup #42782

File: src/metabase/query_processor/middleware/metrics.clj [hunk 3]
```diff
@@ -10,111 +12,160 @@
...
-            new-following-stages (replace-metric-aggregation-refs
-                                  following-stages
-                                  @metric-ref-lookup)
-            combined-stages (vec (remove nil? (concat (butlast pre-transition-stages) [new-metric-stage] new-following-stages)))]
-        (recur query combined-stages metric-ref-lookup))
+  (let [[idx metric-metadata] (find-metric-transition query expanded-stages)
+        [first-stage] expanded-stages]
+    (cond
+      idx
+      (recur query path (update-metric-transition-stages expanded-stages idx metric-metadata))
```
[the remark this answers is not in the document]
Comment (snoe) [reply src/metabase/query_processor/middleware/metrics.clj:134]: Added tests to check this, but metrics based on metrics based on metrics would run into this.

File: src/metabase/query_processor/middleware/metrics.clj (hunk 1 above)
[the remark this answers is not in the document]
Comment (metamben) [reply src/metabase/query_processor/middleware/metrics.clj]: I meant, if it should throw. Maybe the > 1 case shouldn't, but then it should log, I guess.

**Review by metamben — APPROVED** [review]
This is good, any remaining issues can be addressed in a follow up PR.

**Pull request merged**